### DISCUSSION TITLE
Promotion. Castling, and Merging

### DISCUSSION THREAD

**Discussion by rapiermother** [fairy-stockfish/Fairy-Stockfish#657] [Ideas]
I'm developing a GUI for use with Fairy-Stockfish, and I must say that I am thoroughly impressed with it so far. There are a few things I haven't been able to figure out with custom variants (and they may not be doable), so I thought I'd post them here as suggestions.

1. I have a custom piece that's a kind of sub-Bishop. I'd like it to automatically promote to a full Bishop when it reaches the last rank, while Pawns still promote as normal there. The same is desired for a sub-Knight, sub-Rook, and sub-Queen.

2. I have two custom pieces that are a kind of sub-Rook and sub-King. I'd like them to be able to castle with the same rules as the King and Rook (i.e. are on same rank, are on correct files, no intervening pieces, no checks, etc.)

3. I have two custom Pawn pieces with extra movement capability; I would like Pawns and these pieces to progressively promote automatically (Pawn-->PawnLv2-->PawnLv3) _when they capture_. Each of the three Pawn types would have the Pawn's normal last-rank promotion capability as well.

4. I have a custom piece which combines the moves of the Knight and the Queen. I would like for a player's Knight to be able to land in the same square as their own Queen, _merging_ them to form the custom piece.

**Comment by RainRat** [comment]
1. You can set custom promotion paths for each piece. See "promotedPieceType"
2. castlingRookPieces can process a PieceSet, so you can define multiple pieces that are your "Rook". It appears that the other customizations you'd need are hard-coded.
3. There is piecePromotionOnCapture, and as noted, promotedPieceType allows promotion paths for each piece.
4. That is currently not close to being supported.

I am not a primary developer, so there may be inaccuracies.

**Comment by rapiermother** [comment]
Hey, thanks for your help! I'm just focusing on the first item, and I'm having trouble getting the desired result from `promotedPieceType`. My variant is defined this way:

```
[8x10:chess]
pawn = p
knight = n
bishop = b
rook = r
queen = q
king = k
customPiece1 = s:FD
customPiece2 = d:F3
customPiece3 = t:W3
customPiece4 = e:F3W3
customPiece5 = i:FW
maxRank = 10
maxFile = h
startFen = rnbqkbnr/tsdeidst/pppppppp/8/8/8/8/PPPPPPPP/TSDEIDST/RNBQKBNR w KQkq - 0 1
pieceToCharTable = PNBRQ.........SDTEI..Kpnbrq.........sdtei..k
promotionPieceTypes = nbrq
promotionRank = 10
promotedPieceType = s:n d:b t:r e:q
doubleStepRank = 3
pieceValueMg = s:447 d:447 t:702 e:1404 i:574
pieceValueEg = s:459 d:612 t:766 e:1531 i:574
```

If I'm understanding correctly, I'd expect the line `promotedPieceType = s:n d:b t:r e:q` to cause the `s` piece to promote to a Knight when it reaches the back rank; the same for `d` to Bishop, `t` to Rook, and `e` to Queen. None of these promotions happen, though.

My suspicion is that `promotedPieceType` doesn't work the way I expect (and the way implied by @RainRat's comment's inclusion of this information).
